exercises/hangman/.meta/src/reference/java/Hangman.java

```diff
@@ -0,0 +1,89 @@
+import io.reactivex.Observable;
+
+import java.util.*;
+import java.util.stream.IntStream;
+
+import static java.util.stream.Collectors.joining;
+
+class Hangman {
+
+    Observable<Output> play(
+            Observable<String> words,
+            Observable<String> letters) {
+        return Observable
+            .combineLatest(
+                words,
+                letters.startWith(""),
+                (word, letter) -> new AbstractMap.SimpleEntry<>(word, letter))
+            .scan(
+                new Output(
+                    null,
+                    null,
+                    Collections.emptySet(),
+                    Collections.emptySet(),
+                    Collections.emptyList(),
+                    null),
+                (state, entry) -> {
+                    System.out.println(state + " -> " + entry);
+                    if (state == null || state.status != Status.PLAYING) {
+                        return createNewGame(entry.getKey());
+                    } else {
+                        return processNewLetter(state, entry.getValue());
+                    }
+                })
+            .skip(1); // Skip the initial state
+    }
+
+    private static Output createNewGame(String word) {
+        return new Output(
+            word,
+            IntStream.range(0, word.length()).mapToObj(i -> "_")
+                .collect(joining()),
+            new HashSet<>(),
+            new HashSet<>(),
+            new ArrayList<>(),
+            Status.PLAYING);
+    }
+
+    private static Output processNewLetter(
+            Output state,
+            String letter) {
+        if (state.guess.contains(letter) || state.misses.contains(letter)) {
```

Comment:
This type of line should be a method on the `Output` class (or at least a static method) called something like `isAlreadyGuessed`.

Reply:
That's true, and I'll do so.
This is the first complex problem I "implement" for exercism. Do you have guidelines about how much we should help "players" with predefined methods?
I agree that in a real application, `Output` will come with such methods. But being an exercise, I figured that giving the frame of the problem in the form of `Output` without any method was enough to then let them consider the cases to cover. Those cases have corresponding test cases, but it was up to the players to decide how to write the checks.

Reply:
I trust your expertise on this matter so I proceed in adding the methods to `Output`.

Reply:
> This is the first complex problem I "implement" for exercism. Do you have guidelines about how much we should help "players" with predefined methods?

This is a pedagogy question. I think that we should probably only give the students the methods that are used to create the object (and access its state). Our reference solution can add whatever additional methods are useful, but the ones that we provide to the students should not have extra implementation-specific methods.